Feature/apple silicon setup support

sakshamprakash01 opened the pull request:
Created 2 commits-
1. Added a semi-automated script for installing both environments on Apple-Silicon. (credits: 80% GPT-4)
2. Made relevant changes to README.md file

Please advise.

Note: Need to test this thoroughly on an Apple-Silicon Mac.

closes #11

RanFeldesh commented on setup/setup_apple_silicon.py:
```diff
@@ -0,0 +1,69 @@
+def main():
+    platform = identify_operating_system()
+    if platform != 'Darwin':
+        raise Exception('This script is only for macOS with Apple Silicon!')
```
Can we add an automated check for ARM vs. Intel (in addition to `platform != 'Darwin'` that identifies Macs)?

sakshamprakash01 replied:
Resolved by adding paltform.machine() check to verify its "arm64"

RanFeldesh commented on setup/setup_apple_silicon.py:
```diff
@@ -0,0 +1,69 @@
+
+def install_additional_packages(venv_name: str):
+    print(f'Installing additional packages in "{venv_name}" environment...')
+    subprocess.run(f'conda activate {venv_name} && pip install -U numpy', shell=True, check=True)
```
should we have a requirements file (`.txt`) for Mac Arm processors? for uniformity with the `x86` setup

sakshamprakash01 replied:
Added setup/requirements_tensorflow_apple_silicon.txt

RanFeldesh commented on setup/setup_apple_silicon.py:
```diff
@@ -0,0 +1,69 @@
+    install_tensorflow_macos(venv_name)
+    install_additional_packages(venv_name)
+    
+    venv_name = 'pytorch_cpu'
```
👍 
Great that this treats also `PyTorch`, and is using the same process as the `x86` setup

sakshamprakash01 replied:
Yes, this was verified in class but I believe it's a good idea to do an overall final test with the entire script.

RanFeldesh reviewed (commented):
LGTM, but before approving this, let's test it on an M1/M2 system? Let me know how it goes, and if it's fine, let's merge 🚀

sakshamprakash01 commented:
List of changes-
1. Added check for platform i.e. "arm64" to ensure the Mac is using Apple Silicon
2. Renamed "platform" variable to "core_os" as it contains the core os name and not the processor platform name.
3. Added "requirements_tensorflow_apple_silicon.txt" and code changes accordingly.
4. Restructured the code: Moved all utility functions inside a utils.py file.

RanFeldesh commented on setup/README.md:
```diff
@@ -1,4 +1,4 @@
-## Setup
+## Setup (x86 Systems)
```
NIT: every time that we mention `x86` we can right instead `x86/x64`. And clarify that these are Intel and AMD CPUs. 

Context on naming from ChatGPT:
>Intel and AMD CPUs both implement the x86 instruction set architecture (ISA) for 32-bit processors and the x86-64 (also known as AMD64 or x64) instruction set for 64-bit processors. This means that software written for x86 or x86-64 can run on both Intel and AMD CPUs.

>It's worth noting that while both Intel and AMD CPUs share these base instruction sets, each company has also developed its own extensions and variations. For example, Intel has extensions like MMX, SSE, and AVX, and AMD has extensions like 3DNow! and XOP. However, many of these extensions have been adopted by the other company over time, maintaining a high degree of compatibility between the two.

>The shared instruction sets between Intel and AMD processors have played a significant role in their dominance of the desktop, laptop, and server markets, as it allows software developers to write programs that can run on a wide variety of hardware.

Context from ChatGOT about 32 bit vs 64 bit systems:
>The terms 32-bit and 64-bit in the context of processors or operating systems refer to the amount of data that can be processed or the amount of memory that can be addressed at once.

>32-bit (x86): A 32-bit processor can handle 32 bits of data in one processing cycle and can directly address 2^32 (around 4 billion) unique memory locations, which limits the total system memory to 4GB.

>64-bit (x64): A 64-bit processor can handle 64 bits of data in one processing cycle and can directly address 2^64 memory locations. This allows for vastly more system memory (theoretically 18.4 million TB), enabling more efficient processing for tasks that handle large data sets.

>It's also worth noting that 64-bit systems can run both 32-bit and 64-bit software, whereas 32-bit systems are limited to 32-bit software.

Context from ChatGPT about Intel and AMD vs. ARM:
>The primary distinction between Intel/AMD and ARM processors lies in their architecture design philosophy: CISC (Complex Instruction Set Computing) versus RISC (Reduced Instruction Set Computing).

>Intel and AMD: Both Intel and AMD processors are based on the x86 and x64 architectures, which follow the CISC design philosophy. This means they have a wide variety of complex instructions at their disposal, aiming to complete tasks in as few lines of assembly as possible.

>ARM: ARM processors, on the other hand, utilize the RISC design philosophy, which simplifies the processor by efficiently implementing only the instructions that are frequently used in programs. This results in a different trade-off between complexity, power efficiency, and performance compared to CISC designs.
`x86` is 32bit 
>Note that despite these differences, the gap between RISC and CISC has blurred over the years, with modern processors of both types using techniques traditionally associated with the other.

sakshamprakash01 replied:
Got it. Thanks for the context. Certainly it's more appropriate to clarify.

RanFeldesh commented on setup/requirements_tensorflow_apple_silicon.txt:
```diff
@@ -0,0 +1,7 @@
+tensorflow-macos==2.12.0
```
👍 having this list as a separate and dedicate requirements file for Apple Silicon is great (given that the packages are different than those used for `x86/x64` CPUs).

sakshamprakash01 replied:
indeed

RanFeldesh commented on setup/setup_apple_silicon.py:
```diff
@@ -0,0 +1,27 @@
+import utils
```
hmm, would it be better to have a single `setup.py` file, to include all options? that is to have `def main` below inside  the original `setup.py`, and call it from there? Then there won't be much need for a separate `utils.py` file, and the changes you propose here will be minimal and the result most elegant. That is at least the goal, what do you think? would that be indeed the case?

sakshamprakash01 replied:
Now that we can clearly identify the platforms running these scripts, I agree a single setup file will be simpler. Sure, will make changes accordingly.

RanFeldesh replied:
Yes, this was made thanks to you - in identifying the CPU architecture type ! 🚀

sakshamprakash01 commented on setup/setup.py:
```diff
@@ -1,148 +1,26 @@
-    is_git_installed()
-    is_pip_installed()
-    is_ensurepip_installed()
+    core_os = utils.identify_operating_system()
```
I agree that operating_system is a more appropriate term. I will make the necessary changes accordingly.

sakshamprakash01 commented:
The following updates have been made-
1. Re-combined the setup files into a single `setup.py` file catering to all platforms/architectures.
2. Renamed the variable `core_os` to `operating_system`.
3. Made changes to `README.md` file accordingly
4. Formatted all the code in `autopep8` format.

Reminder: Still need to test on ARM64 Macs. Feel free to comment.

RanFeldesh commented on setup/setup.py:
```diff
@@ -1,16 +1,24 @@
 
-def idnetify_shell(platform:str):
+def identify_processor_platform():
+    return platform.machine()
```
NIT: can add a comment with examples of the strings returned

sakshamprakash01 replied:
Added docstrings for every function. LMK if it looks good.

RanFeldesh commented on setup/setup.py:
```diff
@@ -124,25 +122,82 @@ def install_ipykernel(venv_path:str, venv_name:str, platform:str, shell:str, exe
+    print(f'Miniconda environment "{venv_name}" created')
+
+
+def install_tensorflow_deps(venv_name: str):
```
NIT: this is specific to Apple, might want to reflect this in the name of the function

sakshamprakash01 replied:
Renamed it to "install_tensorflow_deps_apple_silicon" and added it to docstrings too.

RanFeldesh commented on setup/setup.py:
```diff
@@ -124,25 +122,82 @@ def install_ipykernel(venv_path:str, venv_name:str, platform:str, shell:str, exe
+        print(f"\n (2/2) Installation Succesful! Please run \'conda activate {venv_name}\' to activate the environment.")
+
+    else:
+        raise Exception('Could not identify the OS/ CPU Architecture of the machine!')
```
NIT: If you want a clearer error message you can write something like: 
```python
f"Could not find a compatible operating system (OS) and CPU Architecture: the platform seems to have OS={operating_system} and CPU architecture={processor_platform}. "
"This setup supports only Windows, Linux or Darwin (Mac) as an OS and x86_64 or arm64 for the CPU architecture."
```

sakshamprakash01 replied:
Added

RanFeldesh commented on setup/setup.py:
```diff
@@ -124,25 +122,82 @@ def install_ipykernel(venv_path:str, venv_name:str, platform:str, shell:str, exe
+    processor_platform = identify_processor_platform()
+
+    # Installation for Windows, Linux, and MacOS running on x86_64 architecture
+    if operating_system in ['Windows', 'Linux', 'Darwin'] and processor_platform == 'x86_64':
```
I guess that the code will work also for `x64` (I _guess_ that this the value returned for AMD, not sure). If so, can add it here.

sakshamprakash01 replied:
Bing on Precise mode gave me-
![image](https://github.com/AASD4011-2023-spring/aasd4011-2023-spring/assets/32038857/4f3cecfb-14aa-4aa3-b089-e207933d6ddf)

Made changes accordingly.

sakshamprakash01 replied:
Will test it before merging.

RanFeldesh reviewed (approved):
LGTM,

Approved, but see comments, and let's test it on a few systems first (one Intel/AMD; one M1/M2 maybe) before merging if possible. 

Great step forward !

sakshamprakash01 commented:
Apart from the above mentioned suggestions,
GPT-4 suggested these improvements in the code (already done):

---

1. In the `validate_python_version()` function, the logical operator `|` should be replaced with `and` to check if the Python version is within the recommended range:

```python
if (sys.version_info >= (3, 7)) and (sys.version_info <= (3, 11)):
```

2. In the `identify_python_executable()` function, there's a typo in the assignment of `executable` variable. It should be `executable = 'python3'` instead of `exectuable = 'python3'`:

```python
elif (executable.endswith('python3')) | (executable.endswith('python3.exe')):
    executable = 'python3'
```

3. Use os.path.join() for `venv_paths` and `requirements_paths` for better compatibility across different operating systems:

```python
venv_paths = [os.path.join('.', 'venv', n) for n in venv_names]
requirements_paths = [os.path.join('.', 'setup', 'requirements_' + e + '.txt') for e in venv_names]
```

sakshamprakash01 commented:
Tested on M1 Macbook Air
---
For some reason, all the macbooks I've used so far (students, friends) had anaconda3 installed. None of them remembered installing anaconda3, however the internet mentions anaconda3 is NOT pre-installed on Macbooks.

Current critical issue: Anaconda3 has only recently been made available with native arm support for apple-silicon (released May 6th). Prior to this, it used to run on rosetta (x86_64 emulation layer). Most of the students have older anaconda3 versions which means running `platform.machine()` outputs `x86_64` even when running on `arm64` machine.

As a solution: Either the students deactivate the `(base)` environment before running `setup.py` or else upgrade to the latest anaconda3 version before running it.

For now, I've added a check to verify the architecture information before installing anything, relying on human feedback.

RanFeldesh commented:
Thank you for the recent update @sakshamprakash01  !

This is great information (and very clear), and it is great that you noticed this in advance !

Happy for your thoughts,  maybe this is an option to consider:
1. Write `setup.py` such that it assumes that either anaconda is deactivated or that it has the new version. So that it assumes that for M1/M2 we do get the `ARM` processor, not `x86_64`.
2. Add a comment in the setup README, in the M1/M2 section, asking the students to either deactivate or install the right version of anaconda. And mention that if they don't do so the setup will fail.

Over time, all students will have the new anaconda, so this will be addressed by default, and the `setup.py` code will stay simple and concise. 

What are your thoughts?

sakshamprakash01 commented:
Found a good solution to check for rosetta (thanks GPT-4)
> You can use the `platform` module in Python to detect the system architecture. However, since Rosetta translates the code for Apple Silicon, it will report the architecture as `x86_64`. To work around this, you can check for the existence of an environment variable `sysctl.proc_translated` that is set to `1` when running under Rosetta.

Updates the changes and tested on M1 Macbook Air. 
Also tested it on my windows and Linux install as well.
Everything works, we are good to merge! 🚀

RanFeldesh reviewed (approved):
LGTM 🚀 
Thank you @sakshamprakash01 ! Good to merge on my end.
